### PR TITLE
Initial Implementation of Height Energy in SGM

### PR DESCRIPTION
**Pull request by crecine**
### Summary

Creation of SGM phases for HeightEnergy, DetailedTakeoff, and DetailedLanding

### Related Issues

- Resolves #

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

**Review by johnjasa — APPROVED** [review]
Good stuff Carl, thanks for this!

Do you plan to add HeightEnergy based tests as part of this PR or a follow-on PR?

**Comment by crecine** [comment]
> Good stuff Carl, thanks for this!
> 
> Do you plan to add HeightEnergy based tests as part of this PR or a follow-on PR?

I can move the unofficial example from the bottom of flops_based/phases/time_integration_phases.py to an official test.

File: aviary/mission/flops_based/ode/mission_ODE.py [hunk 1]
```diff
@@ -169,9 +185,42 @@ def setup(self):
...
+        self.set_input_defaults(Aircraft.Wing.THICKNESS_TO_CHORD,
+                                val=1, units='unitless')
+        self.set_input_defaults(Aircraft.Wing.SWEEP, val=1, units='deg')
+        self.set_input_defaults(Aircraft.Wing.ASPECT_RATIO, val=1, units='unitless')
+        self.set_input_defaults(
+            Aircraft.Design.LIFT_DEPENDENT_DRAG_COEFF_FACTOR, val=1, units='unitless')
+        self.set_input_defaults(
+            Aircraft.Design.ZERO_LIFT_DRAG_COEFF_FACTOR, val=1, units='unitless')
+
+        self.set_input_defaults(Dynamic.Mission.MACH, val=np.ones(nn), units='unitless')
```
[the remark this answers is not in the document]
Comment (crecine) [reply aviary/mission/flops_based/ode/mission_ODE.py:218]: I'll take another look at this

**Comment by johnjasa** [comment]
> > Good stuff Carl, thanks for this!
> > Do you plan to add HeightEnergy based tests as part of this PR or a follow-on PR?
> 
> I can move the unofficial example from the bottom of flops_based/phases/time_integration_phases.py to an official test.

Awesome, please do that!

File: aviary/mission/flops_based/ode/mission_ODE.py [hunk 2]
```diff
@@ -198,11 +217,43 @@ def setup(self):
                            ],
                            promotes_outputs=['initial_mass_residual'])
 
+        if analysis_scheme is AnalysisScheme.SHOOTING:
+            SGM_required_outputs = {
+                Dynamic.Mission.ALTITUDE_RATE: {'units': 'm/s'},
+            }
+            add_SGM_required_outputs(self, SGM_required_outputs)
+
+        print_level = 0 if analysis_scheme is AnalysisScheme.SHOOTING else 2
```
Comment (Kenneth-T-Moore) [inline aviary/mission/flops_based/ode/mission_ODE.py:226]: Kind of wonder why the default is 2 for colocation.

File: aviary/mission/gasp_based/ode/time_integration_base_classes.py [hunk 3]
```diff
@@ -7,23 +7,53 @@
 
 from aviary.mission.gasp_based.ode.params import ParamPort
 from aviary.variable_info.enums import Verbosity
+from aviary.variable_info.variable_meta_data import _MetaData
 
 
 def add_SGM_required_inputs(group: om.Group, inputs_to_add: dict):
-    blank_execcomp = om.ExecComp()
```
Comment (Kenneth-T-Moore) [inline aviary/mission/gasp_based/ode/time_integration_base_classes.py:13]: What is the ultimate purpose of creating a "sink" component for these vars?

File: aviary/mission/gasp_based/ode/time_integration_base_classes.py [hunk 4]
```diff
@@ -7,23 +7,53 @@
...
                         promotes_inputs=list(inputs_to_add.keys()),
                         )
 
 
-# Subproblem used as a basis for forward in time integration phases.
+def add_SGM_required_outputs(group: om.Group, outputs_to_add: dict):
+    iv_comp = om.IndepVarComp()
+    for output, details in outputs_to_add.items():
+        iv_comp.add_output(output, **details)
+    group.add_subsystem('SGM_required_outputs',
```
Comment (Kenneth-T-Moore) [inline aviary/mission/gasp_based/ode/time_integration_base_classes.py:27]: Similar question for this source indepvarcomp.

File: aviary/mission/flops_based/ode/mission_ODE.py [hunk 5]
```diff
@@ -198,11 +217,43 @@ def setup(self):
...
+
+        if analysis_scheme is AnalysisScheme.SHOOTING and False:
```
[the remark this answers is not in the document]
Comment (crecine) [reply aviary/mission/flops_based/ode/mission_ODE.py:228]: It prints out the value of various variables as the trajectory is integrated, it is useful for debugging phases if they fail partway through. I've removed it.

File: aviary/mission/flops_based/ode/mission_ODE.py (hunk 2 above)
Comment (crecine) [reply aviary/mission/flops_based/ode/mission_ODE.py:226]: that's what it was on [line 208](https://github.com/OpenMDAO/Aviary/pull/97/files#diff-d6e2c364a0164fec8fd278a17c15194b88f0b576c84ba74607669322957660d6L208), so I didn't want to change it.

File: aviary/mission/gasp_based/ode/time_integration_base_classes.py [hunk 6]
```diff
@@ -729,7 +811,7 @@ def compute_partials(self, inputs, J):
 
                             dh_j_dx = prob.compute_totals(
                                 [state_name],
-                                prob.states.keys(),
+                                list(prob.states.keys()),
```
Comment (Kenneth-T-Moore) [inline aviary/mission/gasp_based/ode/time_integration_base_classes.py:814]: You should probably just turn these into a list once and save it to a local var. This is several loops deeper from the loop over probs.

File: aviary/mission/gasp_based/ode/time_integration_base_classes.py (hunk 3 above)
Comment (crecine) [reply aviary/mission/gasp_based/ode/time_integration_base_classes.py:13]: the simupy integrator expects that all states be an input to the ODE and all state rates be an output from the ODE. If a state is not directly used in the EOM calculations, this lets you add it without changing the EOM.

File: aviary/mission/gasp_based/ode/time_integration_base_classes.py (hunk 4 above)
Comment (crecine) [reply aviary/mission/gasp_based/ode/time_integration_base_classes.py:27]: same as above, but for the state rates that are not directly calculated by the EOM

**Review by Kenneth-T-Moore — APPROVED** [review]

**Pull request merged**